### ISSUE TITLE
Profiling discussion

### ISSUE THREAD

**Issue by cjolivier01** [apache/incubator-mxnet#9545]
Openning a thread to discuss some items that came up in my Profiling PR, what aren;t directly related to the PR and may go on beyond the life of the PR  

https://github.com/apache/incubator-mxnet/pull/8972

**Comment by TaoLv** [comment]
Hi Chris, thank you for creating this discussion issue. Some questions from my side:

1. Can I get some statistical  infomation from profiling results, like what I did in my naive tool:

    * by op: total time consumed by conv/fc/pooling ...
    * by layer: total time consumed by RNN/LSTM (maybe there is no layer in mxnet)
    * by iteration: total time/memory consumed by one iteration

2. Which is the most time consuming op instance? 

    * input shape, parameters, output shape
    * in which iteration?

3. How to profile memory for CPU?

    * maybe not all memory is allocated in mxnet::storage.

4. Operator tuning: I saw you have contributed to the operator tuning of mxnet. Seems it will create many omp threads before executing graph. How does this function affect the performance of mxnet? If I have set cpu affinity in my environment, these omp threads will be binded to each core. Then when executing computation graph, many other omp threads will be created and binded again. Do you think that will impact the performance?

Thanks.

**Comment by pengzhao-intel** [comment]
Thanks again, @cjolivier01. 
@TaoLv provides very informative questions.

FYI, two points I'd like to mention:

- **MKL-DNN profile information**
When the new CPU backend is enabled from 1.0.1, most of the execution time will be inside MKL-DNN library such as convolution, batch normalization. And we can get the execution time of MKL-DNN by setting MKLDNN_VERBOSE=1, [details](https://github.com/01org/mkl-dnn/blob/master/doc/perf_profile.md). 
The output includes lots of very useful information for the performance debug, one example as below:
 >mkldnn_verbose,exec,convolution,jit:avx512_common,forward_inference,fsrc:nChw16c fwei:OIhw16i16o fbia:undef fdst:nChw16c,alg:convolution_direct,mb32_g1ic64oc64_ih56oh56kh3sh1dh0ph1_iw56ow56kw3sw1dw0pw1,**1.14307**

avx512_common: the executed instruction set architecture
forward_inference: the convolution type
fsrc/fwei/fbia/fdst: format of input data/weight/bias/destination
algorithm: the convolution algorithm, direct convolution
mb32_g1ic64oc64_ih56oh56kh3sh1dh0ph1_iw56ow56kw3sw1dw0pw1: 
  mb, mini-batch, 32
  g, group, 1
  ic, input channel, 64
  oc, ouput channel, 64,
  ih, input height, 56,
  ....
  1.14307, the runtime of this kernel. 

So, you can see the output are really meaningful.
I think we can simplify the MKL-DNN output from MXNET side and make it more readable rather than the raw data.

- **Text Output**
The text level summary may be not such fancy as the visualization tools. But it's really concise and efficient. The user doesn't need to bother chrome and post-processing json file.

A very good example is from Theano's profiling tools [example](http://deeplearning.net/software/theano/tutorial/profiling.html).

As @TaoLv mentioned, this profile includes the runtime from Class, OPs and their instance (Apply).
I really like this kind of format :) 

```
 Function profiling
==================
  Message: None
  Time in 1 calls to Function.__call__: 5.698204e-05s
  Time in Function.fn.__call__: 1.192093e-05s (20.921%)
  Time in thunks: 6.198883e-06s (10.879%)
  Total compile time: 3.642474e+00s
    Theano Optimizer time: 7.326508e-02s
       Theano validate time: 3.712177e-04s
    Theano Linker time (includes C, CUDA code generation/compiling): 9.584920e-01s

Class
---
<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Class name>
  100.0%   100.0%       0.000s       2.07e-06s     C        3        3   <class 'theano.tensor.elemwise.Elemwise'>
   ... (remaining 0 Classes account for   0.00%(0.00s) of the runtime)

Ops
---
<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Op name>
  65.4%    65.4%       0.000s       2.03e-06s     C        2        2   Elemwise{add,no_inplace}
  34.6%   100.0%       0.000s       2.15e-06s     C        1        1   Elemwise{mul,no_inplace}
   ... (remaining 0 Ops account for   0.00%(0.00s) of the runtime)

Apply
------
<% time> <sum %> <apply time> <time per call> <#call> <id> <Apply name>
  50.0%    50.0%       0.000s       3.10e-06s      1     0   Elemwise{add,no_inplace}(x, y)
  34.6%    84.6%       0.000s       2.15e-06s      1     2   Elemwise{mul,no_inplace}(TensorConstant{(1,) of 2.0}, Elemwise{add,no_inplace}.0)
  15.4%   100.0%       0.000s       9.54e-07s      1     1   Elemwise{add,no_inplace}(Elemwise{add,no_inplace}.0, z)
   ... (remaining 0 Apply instances account for 0.00%(0.00s) of the runtime)
```

**Comment by cjolivier01** [comment]
@TaoLv Some aggregate output added today (see screenshot in bottom of PR)
Not all of that yet...

**Comment by cjolivier01** [comment]
You can dump those stats and reset every iteration, I suppose...

**Comment by cjolivier01** [comment]
![screenshot from 2018-01-25 15-55-15](https://user-images.githubusercontent.com/11234557/35419241-b678695e-01eb-11e8-8907-abd3f6ff57f7.png)

**Comment by cjolivier01** [comment]
Like this?

**Comment by cjolivier01** [comment]
I am not yet sure how to incorporate the other MKL output or where I would get it at runtime?

**Comment by cjolivier01** [comment]
Question:

Operator tuning: I saw you have contributed to the operator tuning of mxnet. Seems it will create many omp threads before executing graph. How does this function affect the performance of mxnet? If I have set cpu affinity in my environment, these omp threads will be binded to each core. Then when executing computation graph, many other omp threads will be created and binded again. Do you think that will impact the performance?


So yes, the graph executor itself will be a couple of threads that execute the operators which tend to, in turn, fork into many OMP threads for some parts of the operation (in some cases).  Operator tuning actually supresses this for some cases where the overhead of forking and converging the OMP threads takes longer than it would have taken on a single core.  

But anyway, your question is about does the execution thread binded to a core cause the binded OMP thread to be slower due to caching concerns?  I'd answer "maybe, but without data to the contrary, I assume that since the execution thread tends to be stalled while the operator is running its omp threads, that it's worse to reserve a whole core that will be idle while N-1 OMP threads run".

^^ this assumes multiple operators are not running in parallel, which is a whole other discussion as it relates to core binding.

I am by no means steadfast in this opinion, but it is an assumption at this point.  Have you known this to be the case with other frameworks or experiments?

**Comment by pengzhao-intel** [comment]
@cjolivier01 yes, looks good :)
One minor suggestion is to show the different instance of the Class/OP because their performance will be very different. I think you already have these data :)

Attached the Theano profiling file [theano-profile.log](https://github.com/apache/incubator-mxnet/files/1666204/theano-profile.log).

In the class level, the total runtime is added, such as convoluiton:

```
<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Class name>
15.8%    80.4%      64.328s       3.06e-02s     C     2100       3   theano.sandbox.mkl.mkl_conv.Conv2D
```

In the OP level, the convolution is separated by different shape and input parameters:

```
<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Op name>
10.0%    61.0%      40.886s       5.84e-02s     C      700        1   Conv2D{imshp=(None, None, None, None), kshp=(64, 128, 1, 16), border_mode='valid', subsample=(1, 1), filter_flip=True, filter_dilation=(1, 1)}
 4.7%    76.5%      19.297s       2.76e-02s     C      700        1   Conv2D{imshp=(None, None, None, None), kshp=(128, 10, 16, 1), border_mode='valid', subsample=(1, 1), filter_flip=True, filter_dilation=(1, 1)} 
 1.0%    97.5%       4.145s       5.92e-03s     C      700        1   Conv2D{imshp=(None, None, None, None), kshp=(64, 64, 1, 16), border_mode='valid', subsample=(1, 1), filter_flip=True, filter_dilation=(1, 1)}
```

Then, Theano also gives the runtime and inputs for each instance of convolution OPs.
```
<% time> <sum %> <apply time> <time per call> <#call> <id> <Apply name>
   9.3%    56.5%      40.886s       5.84e-02s    700    39   
Conv2D{imshp=(None, None, None, None), kshp=(64, 128, 1, 16), border_mode='valid', subsample=(1, 1), filter_flip=True, filter_dilation=(1, 1)}
(Relu{slope=0.0}.0, convolution2d_2_W, convolution2d_2_b)
```

Regarding MKL output, we can take a look how to add it to your profile reports when new MKL-DNN backend merged.

**Comment by cjolivier01** [comment]
so far i’ve tried to make the profiling have as close to zero impact as
possible on the runtime performance. i’ll need to think about how to get
those other characteristics in both an elegant, abstract way, while still
keeping the performance impact low (ie avoiding the schrodinger’s cat
problem).

**Comment by pengzhao-intel** [comment]
Yes, it's very important! 
From my experience of Theano profiling, the overhead is very small and maybe you can take a look.

**Comment by cjolivier01** [comment]
can you explicitly list the attributes here please?

**Comment by TaoLv** [comment]
Hi Chris, thanks for your detailed explaination for the omp threads in operator. I used vtune to see the running state of all omp threads in a mxnet workload. I just list my observation here for your information to see if them make sense. I found that 112 omp worker threads were created on my machine (skylake with 56 cores on 2 socket, `OMP_NUM_THREADS` is set to 56). It seems the first 56 threads were created for operator tuning and went into idle after finishing the tuning task. The last 56 threads were created for operator computation. After set `USE_OPERATOR_TUNING=0` and rebuilt mxnet, the first 56 threads were not there anymore. Do you think this behaviour is as expectaton?
Regarding to the omp threads in operator, to my understanding, if there is no operator parallelism, maybe we should give as many threads as possible to the running operator, especially for mkldnn operator. Yes, operator parallelism is totally another topic. In that case, framework should decied how many threads should be given to each operator, so they can be well overlaped.
I think, for mordern deep learning workloads and stencil computation, it would be better to keep a worker thread pool and bind one worker thread to one dedicated core. So the number of worker threads may be same with the number of physical cores on machine. If there are a lot of IO tasks in the workload, maybe we need remain several threads and cores for that. Anyway, thread binding will remove the overhead of thread switching and improve the cache efficiency for dense computation. 

FYI, I use below cmd to set thread affinity and thread number on my machine.
```
export KMP_AFFINITY=granularity=fine,noduplicates,compact,1,0
export OMP_NUM_THREADS=56
```

**Comment by cjolivier01** [comment]
"After set USE_OPERATOR_TUNING=0 and rebuilt mxnet, the first 56 threads were not there anymore. Do you think this behaviour is as expectaton?"
WHat do you mean by "not there anymore"?

**Comment by cjolivier01** [comment]
" if there is no operator parallelism, maybe we should give as many threads as possible to the running operator, especially for mkldnn operator"

It is my understanding that this is the current behavior, is it not?  Certainly there is no code to starve threads from mkl operators, no matter what is going on with the tuning code.

**Comment by cjolivier01** [comment]
"I think, for mordern deep learning workloads and stencil computation, it would be better to keep a worker thread pool and bind one worker thread to one dedicated core. So the number of worker threads may be same with the number of physical cores on machine"

What I have noticed with Intel OMP is that for some N number of "virtual cores", the first N/2 OMP threads are allocated on on each "real" core and then a second pass over the cores for the second N/2 threads.
Currently, we set OMP thread count to N/2, so that OMP threads run one-per-core for the first omp parallelism level (although operator parallelism and omp nesting can change this).

Have you seen behaviour inconsistent with this?

**Comment by cjolivier01** [comment]
Note OpenMP class in src/engine/openmp.cc
```cpp
OpenMP::OpenMP()
  : omp_num_threads_set_in_environment(is_env_set("OMP_NUM_THREADS")) {
#ifdef _OPENMP
  const int max = dmlc::GetEnv("MXNET_OMP_MAX_THREADS", INT_MIN);
  if (max != INT_MIN) {
    omp_thread_max_ = max;
  } else {
    if (!omp_num_threads_set_in_environment) {
      omp_thread_max_ = omp_get_num_procs();
#ifdef ARCH_IS_INTEL_X86
      omp_thread_max_ >>= 1;
#endif
      omp_set_num_threads(omp_thread_max_);
    } else {
      omp_thread_max_ = omp_get_max_threads();
    }
  }
#else
  enabled_ = false;
  omp_thread_max_ = 1;
#endif
}
```

**Comment by TaoLv** [comment]
> "After set USE_OPERATOR_TUNING=0 and rebuilt mxnet, the first 56 threads were not there anymore. Do you think this behaviour is as expectaton?"
> WHat do you mean by "not there anymore"?
I mean just 56 omp worker threads were created and used for operator computation, not for tuning.

**Comment by TaoLv** [comment]
> " if there is no operator parallelism, maybe we should give as many threads as possible to the running operator, especially for mkldnn operator"
> 
> It is my understanding that this is the current behavior, is it not? Certainly there is no code to starve threads from mkl operators, no matter what is going on with the tuning code.

Sorry, I am not very familiar with the current behavior of mxnet. If so, that would be great. :) Can you tell me where in code mxnet decieds how many threads should be given to an operator, or just by `omp_thread_max_`.

**Comment by TaoLv** [comment]
> "I think, for mordern deep learning workloads and stencil computation, it would be better to keep a worker thread pool and bind one worker thread to one dedicated core. So the number of worker threads may be same with the number of physical cores on machine"
> 
> What I have noticed with Intel OMP is that for some N number of "virtual cores", the first N/2 OMP threads are allocated on on each "real" core and then a second pass over the cores for the second N/2 threads.
> Currently, we set OMP thread count to N/2, so that OMP threads run one-per-core for the first omp parallelism level (although operator parallelism and omp nesting can change this).
> 
> Have you seen behaviour inconsistent with this?

I think it's mainly about hyperthreading. If it's on, one physical core will be reported as two hyperthreaded cores. So maybe `omp_thread_max_ >>= 1;` in the code doesn't make sense when hyperthreading is turned off.

Although you can create one omp thread for one core with N/2 when hyperthreading is on. They will still be scheduled by OS and may be switched to different cores if they are not binded to one specific core. That will hurt the performance.

**Comment by cjolivier01** [comment]
If you have better code to determine the number of "real" cores, I will use it.
Right now, we use Intel OMP (at least when building with cmake on Linux) and it reports tat it has bound the first N/2 threads to the respective "real" cores.

**Comment by cjolivier01** [comment]
I added a place for associated performance-related JIRA tasks: 

https://issues.apache.org/jira/secure/RapidBoard.jspa?rapidView=211&projectKey=MXNET&view=planning.nodetail&epics=visible&selectedEpic=MXNET-10

**Comment by TaoLv** [comment]
> If you have better code to determine the number of "real" cores, I will use it.
> Right now, we use Intel OMP (at least when building with cmake on Linux) and it reports tat it has bound the first N/2 threads to the respective "real" cores.

OK, I will look into it. I will submit a new PR after finish that.

**Comment by pengzhao-intel** [comment]
@cjolivier01 I come back to this discussion regarding MKL-DNN backend profiling :)

Now, adding the env "export MKLDNN_VERBOSE =1" and run any CNN benchmark with MXNET master code. You will see the MKL-DNN profiler output.

The C level API is available in here:
https://github.com/intel/mkl-dnn/blob/464c268e544bae26f9b85a2acb9122c766a4c396/include/mkldnn.h#L996

Let's brainstorm how to make a clear profile output for MKL-DNN backend.
